GrpclbLoadBalancerTest is flaky

Depending on how fast the balancing() test method is run, it can sometimes mix up the results of the transports.  I have disabled the test in the mean time, with a todo to re enable it. 
```
java.lang.AssertionError: expected same:<Transport for [address=/127.0.0.1:4000, attrs={}]> was not:<Transport for [address=/127.0.0.1:4001, attrs={}]>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotSame(Assert.java:737)
	at org.junit.Assert.assertSame(Assert.java:680)
	at org.junit.Assert.assertSame(Assert.java:691)
	at io.grpc.grpclb.GrpclbLoadBalancerTest.balancing(GrpclbLoadBalancerTest.java:154)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```

--- a/core/src/main/java/io/grpc/internal/BlankFutureProvider.java
+++ b/core/src/main/java/io/grpc/internal/BlankFutureProvider.java
@@ -40,9 +40,9 @@ import com.google.common.util.concurrent.ListenableFuture;
 import com.google.common.util.concurrent.SettableFuture;
 
 import java.util.Collections;
+import java.util.LinkedHashSet;
 import java.util.Set;
 import java.util.concurrent.CancellationException;
-import java.util.concurrent.ConcurrentHashMap;
 
 import javax.annotation.concurrent.NotThreadSafe;
 
@@ -69,7 +69,7 @@ public final class BlankFutureProvider<T> {
    *
    * <p>If the caller gives up on an undone future, it should call either {@code cancel(true)} or
    * {@code cancel(false)} (they have the same effect) on the future to clean up the tracking
-   * information.
+   * information. The future can be cancelled from any thread.
    */
   public ListenableFuture<T> newBlankFuture() {
     final SettableFuture<T> future = SettableFuture.create();
@@ -118,7 +118,8 @@ public final class BlankFutureProvider<T> {
      * Links the blank futures with futures that will be eventually completed.
      *
      * <p>For each blank future, this method calls {@link Supplier#get()} on {@code source} and link
-     * the returned future to the blank future.
+     * the returned future to the blank future. The blank futures are processed in the same order as
+     * they were created.
      */
     public void link(Supplier<ListenableFuture<T>> source) {
       for (final SettableFuture<T> future : futures) {
@@ -150,6 +151,6 @@ public final class BlankFutureProvider<T> {
     // The cancellation may happen asynchronously, e.g., from a deadline timer thread, thus the set
     // must be thread-safe.
     // There is a race between cancelling and fulfilling, but it is benign.
-    return Collections.newSetFromMap(new ConcurrentHashMap<SettableFuture<T>, Boolean>());
+    return Collections.synchronizedSet(new LinkedHashSet<SettableFuture<T>>());
   }
 }

--- a/core/src/test/java/io/grpc/internal/BlankFutureProviderTest.java
+++ b/core/src/test/java/io/grpc/internal/BlankFutureProviderTest.java
@@ -31,8 +31,8 @@
 
 package io.grpc.internal;
 
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertNotEquals;
 import static org.junit.Assert.assertSame;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
@@ -70,9 +70,8 @@ public class BlankFutureProviderTest {
     });
     assertTrue(f1.isDone());
     assertTrue(f2.isDone());
-    assertNotEquals(f1.get(), f2.get());
-    assertTrue(f1.get().startsWith("value"));
-    assertTrue(f2.get().startsWith("value"));
+    assertEquals("value0", f1.get());
+    assertEquals("value1", f2.get());
     assertFalse(f3.isDone());
   }
 
@@ -95,9 +94,8 @@ public class BlankFutureProviderTest {
     for (int i = 0; i < fulfillingFutures.size(); i++) {
       fulfillingFutures.get(i).set("value" + i);
     }
-    assertNotEquals(f1.get(), f2.get());
-    assertTrue(f1.get().startsWith("value"));
-    assertTrue(f2.get().startsWith("value"));
+    assertEquals("value0", f1.get());
+    assertEquals("value1", f2.get());
     assertFalse(f3.isDone());
   }
 
@@ -136,9 +134,8 @@ public class BlankFutureProviderTest {
     } catch (ExecutionException e) {
       e2 = e.getCause();
     }
-    assertNotEquals(e1.getMessage(), e2.getMessage());
-    assertTrue(e1.getMessage().startsWith("simulated"));
-    assertTrue(e2.getMessage().startsWith("simulated"));
+    assertEquals("simulated0", e1.getMessage());
+    assertEquals("simulated1", e2.getMessage());
     assertFalse(f3.isDone());
   }
 

--- a/grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java
+++ b/grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java
@@ -57,7 +57,6 @@ import io.grpc.Status;
 import io.grpc.TransportManager;
 import io.grpc.internal.ClientTransport;
 
-import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
@@ -89,7 +88,6 @@ public class GrpclbLoadBalancerTest {
   private SettableFuture<ClientTransport> lbTransportFuture;
 
   @Test
-  @Ignore("Flaky") //TODO(zhangkun83): fix flakiness
   public void balancing() throws Exception {
     List<ResolvedServerInfo> servers = createResolvedServerInfoList(4000, 4001);
 

